src/content/blog/rust-based-javascript-linters-lack-typed-linting-right-now/index.mdx

```diff
@@ -0,0 +1,151 @@
+
+Traditionally, lint rules in linters such as ESLint only have visibility into one source code file at a time.
+This makes them fast and theoretically cacheable and parallelizable.
+
+[typescript-eslint](https://typescript-eslint.io) introduces the concept of [linting with type information](https://typescript-eslint.io/linting/typed-linting).
+By calling to TypeScript's type checking APIs, lint rules can make much more informed decisions on code based on types informed by potentially any other file in your project.
+
+Type-checked lint rules can be significantly more capable than traditional lint rules.
+For example:
+
+- [`@typescript-eslint/await-thenable`](https://typescript-eslint.io/rules/await-thenable) disallows `await`s unnecessarily used on non-Thenable (non-Promise) values
+- [`@typescript-eslint/no-floating-promises`](https://typescript-eslint.io/rules/no-floating-promises) lets you know if you create a Promise and forget to handle it safely
+- [`@typescript-eslint/no-for-in-array`](https://typescript-eslint.io/rules/no-for-in-array): flags accidental unsafe `for...in` iterations over arrays (instead of `for...of`)
+
+Each of those rules is only practically useful when they can use type information to determine when to report issues.
+Without type information, they wouldn't be able to understand the type of any value imported from another module.
+
+### Type-Checked Linting Performance
+
+The main downside of type-checked linting is performance.
+Lint rules are now blocked on calling to an API such as TypeScript's, which generally need to read all files to see which ones impact types of any other file.
+That means linting performance will often be strictly worse than that of running `tsc` on your entire project.
+
+We're actively working on this in typescript-eslint.
+Our [Performance Troubleshooting](https://typescript-eslint.io/linting/troubleshooting/performance-troubleshooting) docs have some suggestions, and we're very hopeful that our [`EXPERIMENTAL_useProjectService` option](https://typescript-eslint.io/packages/parser/#experimental_useprojectservice) will land as stable in 2024.
+But even if that option works perfectly then type-checked linting will by design be orders of magnitude slower than traditional linting.
+
+Our experience has been that the benefits of type-checked linting almost always far outweigh the drawbacks of slower performance.
+Most of the time, when we've seen projects with slow type-checked linting, the root cause was either a misconfiguration of typescript-eslint or [slow TypeScript types](https://github.com/microsoft/TypeScript/wiki/Performance).
+
+## Rust-Based Linters and Type Checking
+
+**No Rust-based linter has integrated with TypeScript's type checking APIs yet**.
+That means no Rust-based linter is a full replacement for ESLint + typescript-eslint.
+
+I'm not saying you shouldn't use a Rust-based linter: if you don't want any of the type-checked lint rules, then sure, switching over is great.
```

Comment:
Note from Emanuele and Victorien: it's feasible to run both JS-speed and a native speed linter in tandem. 

Example post: https://github.com/withastro/adapters/pull/111#issuecomment-1866059732

Reply:
https://github.com/typescript-eslint/typescript-eslint/issues/8195